File: SolvedInterviewsQuestionsJava/src/arrays/HeapSort.java
package arrays;

public class HeapSort {
	
	private Integer[] array;
	private int heapSize;
	public HeapSort(Integer[] _array){
		super();
		this.array=_array;
		this.setHeapSize(array.length);
	}
	public Integer[] getArray() {
		return array;
	}
	public int getHeapSize() {
		return heapSize;
	}
	public void setHeapSize(int heapSize) {
		this.heapSize = heapSize;
	}
	
	private int parent(int i){
		//return (int)Math.floor((i-1)/2);
		//return (i-1)/2;
		//System.out.println("parent->"+array[i/2]+"-->index="+i/2);
		return i/2;
	}
	private int left(int i){
		//System.out.println("Left CHild->"+"array[2*i]"+"-->index="+2*i);
		return 2*i;//2*i+1;//2*(i+1);//2*i;//i<<1;    //2*i
	}
	private int right(int i){
		//System.out.println("Right Child->"+"array[2*i+1]"+"-->index="+(2*i+1));
		return 2*i+1;//2*i+2;//2*(i+2);//2*i+1; //i<<1+1;   //2*i+1
	}
	public void maxHeapify(int i){
		int leftChild=left(i);
		int rightChild=right(i);
		//int length=this.array.length;
		int largest;
		if (leftChild<getHeapSize() && this.array[leftChild]>this.array[i]){
			largest=leftChild;
		}else{
			largest=i;
		}
		if(rightChild<getHeapSize() && this.array[rightChild]>this.array[largest]){
			largest=rightChild;
		}
		if (largest!=i){
			int temp=this.array[i];
			this.array[i]=array[largest];
			array[largest]=temp;
			maxHeapify(largest);			
			
		}
				
	}
	public void buildMaxHeap(){
		//int len=this.array.length;
		//setHeapSize(len);
		//for(int i=len/2;i>=0;i--)
		for(int i=getHeapSize()/2;i>=0;i--)
		{
			maxHeapify(i);
		}
		//printArray();
	}
	public void sort(){
		buildMaxHeap();
		printArray();
		for(int i=this.array.length-1;i>=1;i--){
			int temp=this.array[0];
			this.array[0]=this.array[i];
			this.array[i]=temp;
			setHeapSize(i-1);
			maxHeapify(0);
			
		}
	}
	public void printArray()
	{
		for(int i=0;i<this.array.length;i++)
		{
			System.out.print(this.array[i]+" ");
			
		}
		System.out.println();
	}

}

File: Set2inC/HeapSort.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
//#define HeapSize 8
//#define left(i) i<<1;
//#define right(i) (i<<1) +1;
int parent(int);
int left(int);
int right(int);
void MaxHeapify(int A[],int i);
void BuildHeap(int []);
void HeapSort(int []);
int HeapSize=8;
int main(){
	int A[]={5,3,17,10,84,19,6,22,9};
	int i;
	printf("Before Max-Heapifying :\n");
	for(i=0;i<sizeof(A)/sizeof(int);i++)
		printf("%d ",A[i]);
	printf("\n");
	//MaxHeapify(A,1);
	BuildHeap(A);
	printf("After Max-Heapifying :\n");
	for(i=0;i<sizeof(A)/sizeof(int);i++)
		printf("%d ",A[i]);
	printf("\n");
	HeapSort(A);
	printf("After HeapSort :\n");
	for(i=0;i<sizeof(A)/sizeof(int);i++)
		printf("%d\t",A[i]);
	printf("\n");
	return 0;
}

int parent(int i){
	return (floor(i/2));
}

int left(int i){
	//return 1;
	//printf("inside left");
	/*if(i == 0) 
		return 1;
	else 
		return 2*i; //i<<1;*/
	return 2*i+1;
}

int right(int i){
	/*if(i == 0) return 2;
	else 
	return (2*i+1); */
	return 2*i+2;
}

void MaxHeapify(int A[],int i){
	int l,r,temp,max;	
	l=left(i);
	r=right(i);
	if(l<=HeapSize && A[i]<A[l])
		max=l;
	else
		max=i;
	if(r<=HeapSize && A[max]<A[r])
		max=r;	
	//else 
	//	max=i;
	if(max!=i){
		temp=A[i];
		A[i]=A[max];
		A[max]=temp;
//	}
	
	MaxHeapify(A,max);
	}	
}

void BuildHeap(int A[]){
	int i;
	for(i=floor(HeapSize/2);i>=0;i--){
	MaxHeapify(A,i);
}
}

void HeapSort(int A[]){
	int i,temp,x;
	BuildHeap(A);
	for(i=HeapSize;i>=1;i--){
		temp=A[i];
		A[i]=A[0];
		A[0]=temp;
		x=HeapSize-1;
		HeapSize=x;
		MaxHeapify(A,0);
		}
		
}

